### ISSUE TITLE
Make it easier for users to understand the reason for scheduling decisions and pending pods

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#20064]
I was debating whether to file separate issues, but they're similar enough that one issue is probably better.

(1) Make it easier for users to understand the reason for scheduling decisions. For example, when a pod is scheduled, attach an annotation to it with something like this (suggested by @brendandburns )

```
Node-1: Priority =  1234 [1000 (zonal spreading) + 200 (machine spreading) + 30 (balanced resources) + 4 (image affinity)]
Node-2: Priority =  1014 [ ... ]
Node-3: Failed predicate XXX
...
Node-N: ...
```

Also possibly add a kubectl porcelain command to easily fetch this information about a pod.

(2) Make it easier for users to understand the reason for pending pods. This is really the same as (1) except all of the nodes will be marked as "Failed predicate XXX" and the pod in question will be pending rather than running.

In both cases it is also useful to present some kind of summary -- it's not possible to extract a useful signal by reading thousands of lines of this information per pod, as you'd have in a large cluster. There are some fairly obvious ways to summarize the information, but I don't want to stifle anyone's creativity by suggesting anything specific here. :)

**Comment by magicwang-cn** [comment]
@HardySimpson

**Comment by davidopp** [comment]
cc/ @mml

**Comment by HardySimpson** [comment]
Should be some kind of log, but more simple

Fail Predict-Not Enough Source[34] : node1, node2, ......
Fail Predict-Port conflict[28]: node12, node13 .....

Top 10 Priority
Node 17:  Priority =  1234 [1000 (zonal spreading) + 200 (machine spreading) + 30 (balanced resources) + 4 (image affinity)]
Node 23:  Priority =  1200 [1000 (zonal spreading) + 200 (machine spreading) + 0 (balanced resources) + 0 (image affinity)]
.........

**Comment by davidopp** [comment]
@kubernetes/huawei

**Comment by HaiyangDING** [comment]
Refactoring the predicates is part of the work in #20204 , please leave your comments there if you have any suggestions. PS: refactoring the predicates is required to implement #12744.

**Comment by timothysc** [comment]
I think an annotation on the pod itself with a simple description should suffice.  On a side note, do we have any mechanics for rev-ing data-structures in the api?   I would imagine in the long haul it would simply be another field.

**Comment by davidopp** [comment]
> do we have any mechanics for rev-ing data-structures in the api

Annotations that are marked alpha (e.g. `scheduler.alpha.kubernetes.io/name`) have the same compatibility guarantees as alpha APIs, i.e. none. So we can move such annotations into fields at any time (this is exactly what we're planning to do with Affinity in 1.3). Once something is in a field of a non-alpha API (e.g. any field we add to PodSpec), my understanding is we can't change it until the next major API version (v2).

**Comment by mml** [comment]
Previous work in #4583

**Comment by mml** [comment]
@davidopp do we need to summarize this over multiple pods?  E.g. if an RC manages 1000 pods, 800 are pending, and 799 are pending for reason A and 1 is pending for reason B, is outputting 800 lines OK?  My guess is it's better than nothing, which is what we have today, and we can add summarization in the next revision if we need to.

**Comment by mml** [comment]
I believe we have a basic approach sketched out.  For starters, we can implement only the "why pending" part of this problem.  It's simpler and it has a very clear use case: users who wonder why their pod isn't scheduled.

The basic idea is to add a summary to the bottom of `kubectl describe pods` based on the existing `FailedScheduling` events.

- [ ] Currently we only know one reason why a pod failed to schedule on each node.  Let's modify the code so that when a pod fails to schedule on *any* node, we evaluate all the predicates and generate a new `FailedSchedulingEverywhere` event concatenating all those reasons.
- [ ] Add "Why pending" section to `kubectl describe` that produces a nicely formatted version of the info in the `FailedSchedulingEverywhere` event.
- [ ] Audit the events currently generated.  Many lack clarity.  For example, `MatchNodeSelector` failures don't say which selectors failed.
- [ ] As an optimization for larger clusters, add a summary to the end of the Why Pending section.  The code to collapse the fit failure reasons should involve some kind of type signature and a list of functions that as input take sets of failure reasons and as outputs create summaries.  The idea is to make it easy to modify the rules we use to collapse/summarize large numbers of fit failures.  The output should look something like `fit failure on 4 nodes: insufficient CPU, MatchNodeSelector
fit failure on 3 nodes: insufficient CPU`
- [ ] For `kubectl describe {rc,replicaset,daemonset}`, when there are pods pending add a section to the bottom saying something like `Pods are pending.  To find out why, run: kubectl describe pod foo, kubectl describe pod bar, ...`

**Comment by HaiyangDING** [comment]
> I believe we have a basic approach sketched out. For starters, we can implement only the "why pending" part of this problem. It's simpler and it has a very clear use case: users who wonder why their pod isn't scheduled. ...

This sounds like a good plan. I will send a PR to implement this. @mml  @davidopp

**Comment by piosz** [comment]
@mksalawa will work on this

cc @fgrzadkowski

**Comment by goltermann** [comment]
@davidopp @mksalawa did we do this one for 1.4?  If not, can you bump to v1.5?

**Comment by piosz** [comment]
@mksalawa did some progress on this (#29272) but she doesn't work on this anymore. Leaving unassigned.

**Comment by jayunit100** [comment]
ill grab this one

**Comment by davidopp** [comment]
@jayunit100 I assume you'll implement https://github.com/kubernetes/kubernetes/issues/20064#issuecomment-212070819 ?

**Comment by jayunit100** [comment]
w/ annotations dying, is this still a valid approach?

**Comment by davidopp** [comment]
I've lost track of the status (pardon the pun) of this issue, but I think it wouldn't be unreasonable to use an alpha field instead of an annotation, if people wanted to do that. The one PR that has been done for this feature (see upthread) doesn't use annotations or field, anyway.

**Comment by jayunit100** [comment]
I think the conclusion for this issue is really that we would ,for first pass, aggregate event summaries  in the scheduler itself and later on see how it would titrate into API modifications ?  so we can punt on thinking about API stuff for now.

**Comment by davidopp** [comment]
@fgrzadkowski / @piosz would you be willing to take over shepherding issue? it seems close enough to monitoring/metrics that it would make sense for you to shepherd it. Personally, I would really love to see what was described here https://github.com/kubernetes/kubernetes/issues/20064#issuecomment-212070819 implemented at some point. I don't think it would take a lot of time. If @jayunit100 or someone else wants to do it, that's great.

**Comment by timothysc** [comment]
We need to clean the events, and I would take a wholistic approach here.  At scale, on a bad affinity you can easily flood events with garbage.  /cc @wojtek-t

**Comment by jayunit100** [comment]
gated on ^ , continue dialog there.

```
add support for scheduling metadata in pod api + gut reasons events ----> Update pod metadata api to support summary info.
```

**Comment by dims** [comment]
This needs to be triaged as a release-blocker or not for 1.5 @jayunit100 @timothysc @davidopp

**Comment by timothysc** [comment]
We've done some cleanup and perf work in 1.5 and we'll move the rest to 1.6

**Comment by timothysc** [comment]
@jayunit100 do you have any state update here now that 1.6 is opening up?

**Comment by jayunit100** [comment]
Yeah: we've cleaned the events stuff into a roll up, that gives a non kubectl way to see a birds eye view summary of all the scheduling failures.  

I'm going to start ramping up on API refactoring tomorrow, that is a longer term task

**Comment by jayunit100** [comment]
TL'DR, here's where we are at, we need 4 PR's.

- [ x ] Making a rollup of events that makes sense... I think that is done.
- [ ] Audit events that lack important details (i.e. node not found, 10+ of these!)
- [ ] API level support for the `SchedulingReasons` data structures
- [ ] Kubectl level support for users

**Comment by timothysc** [comment]
>  API level support for the SchedulingReasons data structures 

Reason ('s') - how many are we looking to store?  I would expect only the last reason is the important one.

**Comment by jayunit100** [comment]
map (node -> reason)

**Comment by jayunit100** [comment]
cc @davidopp  making the Scheduling reasons data structure independent of node specific info, can you update the issue header?

**Comment by jayunit100** [comment]
We explictly are NOT planning to do a follow on w/ node specific details, its not necessary.  the above bullets will close this issue out

**Comment by davidopp** [comment]
Sorry, can you explain which part of https://github.com/kubernetes/kubernetes/issues/20064#issuecomment-212070819 you don't think we should do?

**Comment by jayunit100** [comment]
- We are implementing 20064, 
- We are *not* implementing the original issue description which has node specific details in it., that is , we are doing `map[reason]->count` as opposed to `map[node]->reason` .

**Comment by smarterclayton** [comment]
Comment from the linked PR:

I don't think this would be the only information we ever surface from scheduling.  I would expect in the future to grow the status here about why a schedule fails.  I don't see consensus here about the full scope of this pattern (scheduler writing pod status), so I'd like to see a quick description of that - a single field here isn't the end of that info, and I don't want to piecemeal add these options in the future without some understanding of the overarching design.

I saw the suggestion from mml to use the event to infer this info (which the pod description already has available).  Why does it need to specifically be on pod status?

**Comment by jayunit100** [comment]
To summarize how we got here.

- in an earlier issue https://github.com/kubernetes/kubernetes/issues/35842 we decided we didn't want scheduler events to be ~O(P*N), so we rolled them up.  That could definitely provide user feedback,
- *however* we decided this feature is sufficiently important that we *dont* want to depend on ephemeral data.

hmmm... Regarding the question above, can you clarify what you mean by  `it need to specifically be on pod status` ?  I wasn't sure what `it` is referring to there...?

**Comment by davidopp** [comment]
I'll re-read this and throw in my two cents in the next few days. IIUC the main controversy is whether this informatoin should go in PodStatus?

**Comment by jayunit100** [comment]
i think to summarize clayton's thoughts...  there are 2 minor contraversies:
- should the scheduler updating pod status type information fields? imo podspec is a decent place for it to live also.
- other information should we want to surface from the scheduler ? should we think about it and have it tackled in this API change? imo not necessary for this feature,but i could see the value in doing so.

**Comment by smarterclayton** [comment]
Third bullet:

If this is for users, and we're not totally clear on how this would be used
by automated processes in the future, rather than add to the API we should
add a message or condition which summarizes this information for the end
user in a human friendly way.  That is better than adding an API because
this information would already be exposed by clients and thus reach users
more effectively.

On Jan 5, 2017, at 3:45 AM, David Oppenheimer <notifications@github.com>
wrote:

I'll re-read this and throw in my two cents in the next few days. IIUC the
main controversy is whether this informatoin should go in PodStatus?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/20064#issuecomment-270592235>,
or mute the thread

**Comment by jayunit100** [comment]
So, are we

- erasing and rewriting Message?
- appending to Message and aggregating only in kubectl?
- aggregating the map into Message?

**Comment by smarterclayton** [comment]
The nice thing about scheduling is that you're at the right spot:

```
	// Current service state of pod.
	// More info: http://kubernetes.io/docs/user-guide/pod-states#pod-conditions
	// +optional
	Conditions []PodCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,2,rep,name=conditions"`
	// A human readable message indicating details about why the pod is in this condition.
	// +optional
	Message string `json:"message,omitempty" protobuf:"bytes,3,opt,name=message"`
	// A brief CamelCase message indicating details about why the pod is in this state.
	// e.g. 'OutOfDisk'
	// +optional
	Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
```

You can set a reason `CannotSchedule` (the machine readable reason why you are in this condition), a message `There are no nodes that match the node predicates X, Y, and Z` (human readable) and probably a condition on the pod.  Once we have competitive schedulers this could be more difficult (because they'd be fighting over reason), but I do think it's possible we can come back to that.

**Comment by jayunit100** [comment]
Yes, its definitely not a long term solution but *good... enough* for now.

**Comment by jayunit100** [comment]
yup, something like 

reason = `cannotSchedule` + message `There are no nodes that matched all the predicates:  The map of predicate failures is {"memory":23, "cpu":30}`.

makes sense.

**Comment by smarterclayton** [comment]
Reason is a constant and is always CamelCase "CannotSchedule" .

Please ensure the message is readable by a human.  The map output above
really isn't something I would ever show to a user.

**Comment by jayunit100** [comment]
yup typo fixed

**Comment by smarterclayton** [comment]
Can you double check that we don't already have a condition for "scheduling failed" or "cannot schedule"?  Someone mentioned we discussed it at one point, not sure whether it ever got implemented.

**Comment by jayunit100** [comment]
yup im in the code now, `PodReasonUnschedulable` is a valid pod condition.

**Comment by smarterclayton** [comment]
Ok, so I'd recommend adding the condition with the message, adding a reason on the condition that is more specific (i.e. "NoNodesMatchingPredicates" or something like that), and then using "Unschedulable" as the pod status level reason and the same message as the condition.

```
PodStatus {
  Conditions: []{
    PodCondition{
      Type: "Unschedulable",
      Value: true,
      Reason: "NoNodesMatchingPredicates",
      Message: "No nodes are available that match the following predicates: memory, cpu",
      ... // other fields
    },
  }
  Reason: "Unschedulable",
  Message: "<same as condition>",
}
```

**Comment by jayunit100** [comment]
makes sense,  the PodStatus updater in the scheduler handles the update mutate logic to .

**Comment by davidopp** [comment]
I'm going to reopen until I have a chance to take a look at what you did. Still haven't had time unfortunately.

**Comment by jayunit100** [comment]
@davidopp this is all i did, not a huge change:  feel free to keep open or make new issues as needed.

run something like:
```
[root:...olang/src/k8s.io/kubernetes]# 
cluster/kubectl.sh get pod redis-master-223378846-1k1d8 -o=json | grep -A 10 status                                                                           
```
and get the scheduling failures listed in conditions:
```
(sched_Histogram_error✱)
    "status": {
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-01-06T20:32:28Z",
                "message": "No nodes are available that match all of the following predicates: Insufficient memory (1).",
                "reason": "Unschedulable",
                "status": "False",
                "type": "PodScheduled"
            }
        ],
        "phase": "Pending"
    }
}
```

**Comment by timothysc** [comment]
>  "No nodes are available that match all of the following 
predicates: Insufficient memory (1)."

Instead of a list of events per node, there is a histogram of failures printed in string list on the pod status message.

**Comment by timothysc** [comment]
closing, fixed by #39491
